### ISSUE TITLE
Allow nested transitions

### ISSUE THREAD

**Issue by vanhoofseppe** [viewflow/django-fsm#125]
Now it is possible to use the signal 'post_transition' to handle your things after your state has been changed. But the signal is send out to all of the FSM receivers you configure, and has to filtered out again in each receiver. What will be more intensive if your project is growing.

When you have a child object that will change state you will want to auto check your parent object if all his children are to the wanted state to proceed to the next parent state. But the last child that is still in the previous state is not yet transferred to the new state and has to be saved first, and then we are outside the state machine, and we can not automagically change the parent object. While I think this is still something to be handled inside the state machine.
And I get why you don't want to auto save. But now it is not possible for others to implement a after save method within the FSM.

If there would be a method that is called after the transition, we can implement our own save at that moment without working with signals and receivers, and a lot of more is possible with this FSM.

Thanks,

**Comment by salteggs** [comment]
I agree, signals are bad practice according to Two Scoops of Django.

A more direct approach would avoid possible wrong execution of code and it would help performance-wise if many models utilise the Django FSM.

**Comment by kmmbvnr** [comment]
Generally, you don;t need any fsm for doing after transition

```
class MyModel(db.Models):

     @transition(..)
     def do_x(self):
         ....

     def do_x_y(self):
          self.do_x()
          do_after_transition()
```

**Comment by vanhoofseppe** [comment]
Ow Great! haven't tough like this! Thanks!

**Comment by kmmbvnr** [comment]
But i agree, that same kind of problem happens when we try to call a transition method inside another transition method, ex::

    @transition(source='x', target='y')
    def transition1(self):
         ...
         self.transition2()

    @transition(source='y', target='z')
    def transition2(self):
         ...

**Comment by vanhoofseppe** [comment]
But now we can do:
`
class MyParentModel(db.Models):

     @transition(source='x', target='y')
     def transition1(self):
          ...

     def transition1_function(self):
          self.check_my_child_model()  # what will work because the child is saved
          # the above will be my condition checker
          self.transition1()
          self.save()

class MyChildModel(db.Models):

     @transition(source='x', target='y')
     def transition2(self):
          ...

     def transition2_function(self):
          self.transition2()
          self.save()  # what will fix my problem
          self.parent.transition1_function()
`

**Comment by kmmbvnr** [comment]
The new version of FSM have no such issue b/c the target state value set before entering to the transition method

https://github.com/viewflow/viewflow/blob/master/viewflow/fsm.py

**Comment by jpdavy** [comment]
@kmmbvnr - Love this package. Thanks for sharing it. I am experiencing this nested transition issue too. When I do: 

`    @transition(field=state, source=NOT_STARTED, target=STARTING, conditions=[lambda i: i.total_contacts > 0])
    def start(self):
        from .redis import rq_start_campaign
        django_rq.enqueue(rq_start_campaign, self)
        return self.resume()

    @transition(field=state, source=[STARTING, PAUSED], target=STARTED,)
    def resume(self):
        return
`

Basically I'd like to run some automated code in the first state, and as long as this code executes correctly, transition to the next state automatically. The first state is one-directional, which prevents that code from being run multiple times.

I just installed django-fsm today, and not seeing this fix. Any suggestions?

**Comment by kmmbvnr** [comment]
Implemented in the `viewflow.fsm` package - https://docs-next.viewflow.io/fsm/options.html#target

The stage changed before entering a method body, that allows to to call the next transition method

On exception, state are returned back to the original
